File: backend/src/snowan/server/skills.py
"""Agent skills: list / get / create / update / enable / delete."""
from fastapi import APIRouter, HTTPException
from fastapi.responses import Response
from pydantic import BaseModel

from .. import skills as store

router = APIRouter(prefix="/api/skills")


class SkillCreate(BaseModel):
    name: str
    description: str = ""
    body: str = ""


class SkillUpdate(BaseModel):
    description: str | None = None
    body: str | None = None


class EnableBody(BaseModel):
    enabled: bool


class ImportBody(BaseModel):
    paths: list[str]


@router.get("")
def list_skills() -> list[dict]:
    return store.list_skills()


# Registered before /{name} so "external" isn't captured as a skill name.
@router.get("/external")
def list_external() -> list[dict]:
    return store.discover_external()


@router.post("/import")
def import_external(req: ImportBody) -> dict:
    return {"imported": store.import_external(req.paths)}


@router.get("/{name}")
def get_skill(name: str) -> dict:
    s = store.get_skill(name)
    if s is None:
        raise HTTPException(404, "未找到技能")
    return s


@router.post("")
def create(req: SkillCreate) -> dict:
    if not req.name.strip() or not req.body.strip():
        raise HTTPException(422, "需要名称和内容")
    return store.create_skill(req.name.strip(), req.description, req.body)


@router.put("/{name}")
def update(name: str, req: SkillUpdate) -> dict:
    s = store.update_skill(name, description=req.description, body=req.body)
    if s is None:
        raise HTTPException(404, "未找到技能")
    return s


@router.patch("/{name}/enabled")
def set_enabled(name: str, req: EnableBody) -> dict:
    store.set_enabled(name, req.enabled)
    return {"ok": True}


@router.delete("/{name}")
def remove(name: str) -> Response:
    if not store.delete_skill(name):
        raise HTTPException(404, "未找到技能")
    return Response(status_code=204)

File: frontend/src/api/skills.ts
import { api } from './base';

export type Skill = {
  name: string;
  description: string;
  enabled: boolean;
  tags?: string[];
  source?: string | null;
};
export type SkillDetail = Skill & { body: string };
export type ExternalSkill = {
  name: string;
  description: string;
  source: string;
  path: string;
  already: boolean;
};

const j = <T>(r: Response): Promise<T> => {
  if (!r.ok) throw new Error(`${r.status}`);
  return r.json();
};

const send = <T>(method: string, path: string, body?: unknown): Promise<T> =>
  fetch(api(path), {
    method,
    headers: body ? { 'Content-Type': 'application/json' } : undefined,
    body: body ? JSON.stringify(body) : undefined,
  }).then((r) => j<T>(r));

export const listSkills = () => fetch(api('/api/skills')).then((r) => j<Skill[]>(r));

export const getSkill = (name: string) =>
  fetch(api(`/api/skills/${encodeURIComponent(name)}`)).then((r) => j<SkillDetail>(r));

export const createSkill = (name: string, description: string, body: string) =>
  send<SkillDetail>('POST', '/api/skills', { name, description, body });

export const updateSkill = (name: string, patch: { description?: string; body?: string }) =>
  send<SkillDetail>('PUT', `/api/skills/${encodeURIComponent(name)}`, patch);

export const setEnabled = (name: string, enabled: boolean) =>
  send<{ ok: boolean }>('PATCH', `/api/skills/${encodeURIComponent(name)}/enabled`, { enabled });

export const deleteSkill = (name: string) =>
  fetch(api(`/api/skills/${encodeURIComponent(name)}`), { method: 'DELETE' });

export const discoverExternalSkills = () =>
  fetch(api('/api/skills/external')).then((r) => j<ExternalSkill[]>(r));

export const importSkills = (paths: string[]) =>
  send<{ imported: Skill[] }>('POST', '/api/skills/import', { paths });
